4227 update dev/test env usecase with demo notebooks and walkthrough video (#4277)
treeverse/lakeFS: docs/use_cases/iso_env.md

title: Isolated Dev/Test Environment
description: In this tutorial, we will explore how to use lakeFS to create isolated dev/test data environments to run data pipelines.
parent: Use Cases
nav_order: 10
has_children: false
---

# Isolated Dev/Test Environment

## Why do I need multiple environments?

When working with a data lake, it's useful to have replicas of your production environment. These replicas allow you to test and understand changes to your data without impacting the consumers of the production data.

Running ETL and transformation jobs directly in production is a guaranteed way to have data issues flow into dashboards, ML models, and other consumers sooner or later. The most common approach to avoid making changes directly in production is to create and maintain multiple data environments. Dev environment to develop the data pipelines and test environment where pipeline changes are implemented before pushing it to production.

The issue with this approach is that it's time-consuming and costly to maintain these separate dev/test environments. And for larger teams it forces multiple people to share these environments, requiring significant co-ordination.

> ## How do I create dev/test environments with lakeFS?

Reviewer: "## How do I create an isolated data dev/test environment with lakeFS?

Objects that remain unchanged between two branches are not copied, but rather shared to both branches via metadata pointers that lakeFS manages. If you make a change on one branch and want it reflected on another, you can perform a `merge` operation to update one branch with the changes from another.
{: .note }

Let’s see an example of using multiple lakeFS branches for isolation.

## Using branches as environments

The key difference when using lakeFS for isolated data environments is that you can create them immediately before testing a change. And once new data is merged into production, you can delete the branch - effectively deleting the old environment.

This is different from creating a long-living test environment used as a staging area to test all the updates. With lakeFS, **we create a new branch for each change to production** that we want to make. One benefit of this is the ability to test multiple changes at one time.

![dev/test branches as environments]({{ site.baseurl }}/assets/img/iso_env_dev_test_branching.png)

# Creating Dev/Test Environments with lakeFS

> lakeFS supports UI, CLI (`lakectl` commandline utility) and several client API [integrations](../integrations) to run the git-like operations. Let us explore how to create dev/test environments using each of these options below.

Reviewer 2: ```suggestion
lakeFS supports UI, CLI (`lakectl` commandline utility) and several client API [integrations](../integrations) to run Git-like operations. Let us explore how to create dev/test environments using each of these options below.
```

## Using lakeFS Playground UI
In this tutorial, we will use [lakeFS playground](https://demo.lakefs.io/) to create dev/test data environments. Playground allows you to spin up a lakeFS instance in a click, create different data environments by simply branching out of your data repository and develop & test data pipelines in these isolated branches.

First, let us spin up a [playground](https://demo.lakefs.io/) instance. Once you have a live environment, login to your instance with access and secret keys. Then, you can work with the sample data repository `my-repo` that is created for you.

![sample repository]({{ site.baseurl }}/assets/img/iso_env_myrepo.png)

Click on `my-repo` and notice that by default, the repo has a `main` branch created and `sample_data` preloaded to work with.

![main branch]({{ site.baseurl }}/assets/img/iso_env_sampledata.png)

> You can create a new branch say `test-env` by going to the branches tab and clicking `Create branch`. Once it is successful, we see two branches under the repo: `main` and `test-env`.

Reviewer 2: ```suggestion
You can create a new branch (say, `test-env`) by going to the _Branches_ tab and clicking _Create Branch_. Once it is successful, you will see two branches under the repo: `main` and `test-env`.
```

![test-env branch]({{ site.baseurl }}/assets/img/iso_env_testenv_branch.png)

> Now, you can add, modify or delete the data files in the `test-env` branch without affecting the data in the main branch.

Reviewer 2: ```suggestion
Now you can add, modify or delete objects under the `test-env` branch without affecting the data in the main branch.
```

## Using lakeFS Python Client API and Jupyter notebook

> Adding or deleting files from different branches in a lakeFS repo can be performed from the UI as shown above. However, to perform complex transformations on the data using Spark or other compute engine, you would need to use lakeFS supported APIs.

Reviewer 2: ```suggestion
Adding or deleting objects to different branches in a lakeFS repository can be performed from the UI as shown above. However, to perform complex transformations using Spark or other compute engines, you will need to use lakeFS clients.
```

> This use case shows how you can create dev/test data environments using lakeFS branches. The following tutorial will use an existing lakeFS environment (i.e., playground) and a jupyter notebook, python lakefs_client API to demonstrate integration of lakeFS with [Spark](../integrations/spark). You can run this tutorial on your local machine.

Reviewer 2: Not clear.

Author: Fixed it.

### Prerequisites

To instantly spin up a lakeFS environment and to work with sample data, use [lakeFS playground](https://demo.lakefs.io/). To use lakeFS with your own production data or to run it in your own machine, check out [Quickstart](../quickstart/index.md) guide.

Before getting started, you will need [docker](https://docs.docker.com/engine/install/) installed on your machine.

{: .note }
Once you have docker installed and a lakeFS instance running, you can access the lakeFS demo notebooks by cloning the [lakeFS-samples](https://github.com/treeverse/lakeFS-samples) git repo.

Follow along the steps below to create dev/test environment with lakeFS.

* Start by cloning the repository:
```bash

> git clone https://github.com/treeverse/lakeFS-samples.git && cd 03-apache-spark-python-demo

Reviewer: love the usage of the sample repository!!!

```bash
storageNamespace = 's3://<S3 Bucket Name>/' # e.g. "s3://treeverse-demo-lakefs-storage-production/user_expert-robin/my-repo"
```

You can use lakeFS through the UI, API or `lakectl` commandline. For this use-case, we use python `lakefs_client` to run lakeFS core operations.

```bash
import lakefs_client
from lakefs_client import models
from lakefs_client.client import LakeFSClient

# lakeFS credentials and endpoint
configuration = lakefs_client.Configuration()
configuration.username = lakefsAccessKey
configuration.password = lakefsSecretKey
configuration.host = lakefsEndPoint

client = LakeFSClient(configuration)
```

lakeFS can be configured to work with Spark in two ways:
* Access lakeFS using the [S3A gateway](../integrations/spark.md#access-lakefs-using-the-s3a-gateway)
* Access lakeFS using the [lakeFS-specific Hadoop FileSystem](../integrations/spark.md#access-lakefs-using-the-lakefs-specific-hadoop-filesystem)

### Upload the sample data to main branch

> To upload a file to the `my-repo`, use the following command.

Reviewer 2: ```suggestion
To upload an object to the `my-repo`, use the following command.
```

```bash
import os
contentToUpload = open(os.path.expanduser('~')+'/lakefs_test.csv', 'rb')
client.objects.upload_object(
    repository="my-repo",
    branch="main",
    path=fileName, content=contentToUpload)
```

Once uploaded, commit the changes to the `main` branch and attach some metadata to the commit as well.
```bash
client.commits.commit(
    repository="my-repo",
    branch="main",
    commit_creation=models.CommitCreation(

>         message='Added my first file!',

Reviewer 2: ```suggestion
        message='Added my first object!',
```
